TouchPortalAPI/tppdoc.py

```diff
@@ -0,0 +1,426 @@
+
+    for category in filterConnectorsbyCategory:
+        connectorDoc += f"<details {'open' if list(filterConnectorsbyCategory.keys()).index(category) == 0 else ''}><summary><b>Category:</b> {category} <ins>(Click to expand)</ins></summary>"
+        connectorDoc += filterConnectorsbyCategory[category]
+        connectorDoc += "</table></details>\n"
+    connectorDoc += "<br>\n"
+
+    return connectorDoc
+    
+
+
+def generateSetting(entry):
+    settingDoc = "\n\n## Settings Overview\n"
+
+    def f(data):
+        return [data.get('maxLength', 0) > 0, data.get('minValue', None), data.get('maxValue', None)]
+
+    for setting in entry.keys():
+        settingDoc += f"### {setting}\n"
+        settingDoc += "| Read-only | Type | Default Value"
+        if f(entry[setting])[0]: settingDoc += f" | Max. Length"
+        if f(entry[setting])[1]: settingDoc += f" | Min. Value"
+        if f(entry[setting])[2]: settingDoc += f" | Max. Value"
+        settingDoc += " |\n"
+        settingDoc += "| --- | --- | ---"
+        if f(entry[setting])[0]: settingDoc += f" | ---"
+        if f(entry[setting])[1]: settingDoc += f" | ---"
+        if f(entry[setting])[2]: settingDoc += f" | ---"
+
+        settingDoc += " |\n"
+        settingDoc += f"| {entry[setting]['readOnly']} | {entry[setting]['type']} | {entry[setting]['default']}"
+        if f(entry[setting])[0]: settingDoc += f" | {entry[setting]['maxLength']}"
+        if f(entry[setting])[1]: settingDoc += f" | {entry[setting]['minValue']}"
+        if f(entry[setting])[2]: settingDoc += f" | {entry[setting]['maxValue']}"
+        settingDoc += " |\n\n"
+        if entry[setting].get('doc'):
+            settingDoc += f"{entry[setting]['doc']}\n\n"
+    return settingDoc
+
+def generateState(entry, baseid):
+    stateDoc = "\n## States\n"
+    filterCategory = {}
+    for state in entry:
+        categoryName = entry[state].get("category", "main")
+        if not categoryName in filterCategory:
+            filterCategory[categoryName] = ""
+            filterCategory[categoryName] += f"<details{' open' if len(filterCategory) == 1 else ''}><summary><b>Base Id:</b> {baseid} <b>Category:</b> {entry[state].get('category')} <ins>(Click to expand)</ins></summary>\n"
+            filterCategory[categoryName] += "\n\n| Id | Description | DefaultValue | parentGroup |\n"
+            filterCategory[categoryName] += "| --- | --- | --- | --- |\n"
+
+        filterCategory[categoryName] += f"| {entry[state]['id'].split(baseid)[-1]} | {entry[state]['desc']} | {entry[state]['default']} | {entry[state].get('parentGroup', ' ')} |\n"
+
+    for category in filterCategory:
+        stateDoc += filterCategory[category]
+        stateDoc += "</details>\n\n"
+    stateDoc += "<br>\n"
+
+    return stateDoc
+
+def generateEvent(entry, baseid):
+    eventDoc = "\n## Events\n\n"
+    filterCategory = {}
+
+    for event in entry:
+        needDropdown = False
+
+        categoryName = entry[event].get("category", "main")
```

Comment:
And here (and below) you already have the `event` object from the `for` loop but then you get it again and again from within the `entry` array by using itself as a key...   :-)

Reply:
Sorry. my mistake.